[CC 5965] Improve HCP Observability E2E tests and add periodic tests

Achooo opened the pull request:
Impact of this test in reality is minimal as this test currently **does not run** in CI yet, it is used locally for confidence in our Consul Core changes.

Changes proposed in this PR:
1. Rename the test `TestBasicCloud` to `TestObservabilityCloud` and filename`cloud/test_observability.go`
2. Refactor (cleanup) the test with a fakeserver http client object
3. Add a table driven set of scenarios to the Cloud tests for observability:
- Collector exports metrics
- Consul exports metrics
- Periodic refresh: metrics configuration changes, Consul exports metrics with this new config (new filters and label)
- Periodic refresh : disabled metrics, configuration changes to disable metrics, Consul stops exporting metrics

How I've tested this PR:
- Ran tests locally using custom docker images for both the fakeserver and consul

see below:

```
❯ go test -run TestObservabilityCloud -v -p 1 -timeout 20m \
                -use-kind \
                -consul-collector-image hashicorp/consul-telemetry-collector:0.0.1 -consul-image docker.io/achooo/consul01:17
=== RUN   TestObservabilityCloud
    command.go:100: 2023-09-12T11:12:28-04:00 Running command kubectl with args [--namespace default apply -k ../fixtures/bases/cloud/hcp-mock]
    command.go:179: 2023-09-12T11:12:28-04:00 serviceaccount/fake-server created
    command.go:179: 2023-09-12T11:12:28-04:00 service/fake-server created
    command.go:179: 2023-09-12T11:12:28-04:00 deployment.apps/fake-server created
    command.go:100: 2023-09-12T11:12:28-04:00 Running command kubectl with args [--namespace default kustomize ../fixtures/bases/cloud/hcp-mock]
    command.go:179: 2023-09-12T11:12:28-04:00 apiVersion: v1
    command.go:179: 2023-09-12T11:12:28-04:00 kind: ServiceAccount
    command.go:179: 2023-09-12T11:12:28-04:00 metadata:
    command.go:179: 2023-09-12T11:12:28-04:00   name: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00 ---
    command.go:179: 2023-09-12T11:12:28-04:00 apiVersion: v1
    command.go:179: 2023-09-12T11:12:28-04:00 kind: Service
    command.go:179: 2023-09-12T11:12:28-04:00 metadata:
    command.go:179: 2023-09-12T11:12:28-04:00   name: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00 spec:
    command.go:179: 2023-09-12T11:12:28-04:00   ports:
    command.go:179: 2023-09-12T11:12:28-04:00   - name: https
    command.go:179: 2023-09-12T11:12:28-04:00     port: 443
    command.go:179: 2023-09-12T11:12:28-04:00     targetPort: 443
    command.go:179: 2023-09-12T11:12:28-04:00   - name: http
    command.go:179: 2023-09-12T11:12:28-04:00     port: 8080
    command.go:179: 2023-09-12T11:12:28-04:00     targetPort: 8080
    command.go:179: 2023-09-12T11:12:28-04:00   selector:
    command.go:179: 2023-09-12T11:12:28-04:00     app: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00 ---
    command.go:179: 2023-09-12T11:12:28-04:00 apiVersion: apps/v1
    command.go:179: 2023-09-12T11:12:28-04:00 kind: Deployment
    command.go:179: 2023-09-12T11:12:28-04:00 metadata:
    command.go:179: 2023-09-12T11:12:28-04:00   name: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00 spec:
    command.go:179: 2023-09-12T11:12:28-04:00   replicas: 1
    command.go:179: 2023-09-12T11:12:28-04:00   selector:
    command.go:179: 2023-09-12T11:12:28-04:00     matchLabels:
    command.go:179: 2023-09-12T11:12:28-04:00       app: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00   template:
    command.go:179: 2023-09-12T11:12:28-04:00     metadata:
    command.go:179: 2023-09-12T11:12:28-04:00       labels:
    command.go:179: 2023-09-12T11:12:28-04:00         app: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00       name: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00     spec:
    command.go:179: 2023-09-12T11:12:28-04:00       containers:
    command.go:179: 2023-09-12T11:12:28-04:00       - image: docker.io/achooo/fakeserver:test50
    command.go:179: 2023-09-12T11:12:28-04:00         name: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00         ports:
    command.go:179: 2023-09-12T11:12:28-04:00         - containerPort: 443
    command.go:179: 2023-09-12T11:12:28-04:00           name: https
    command.go:179: 2023-09-12T11:12:28-04:00         - containerPort: 8080
    command.go:179: 2023-09-12T11:12:28-04:00           name: http
    command.go:179: 2023-09-12T11:12:28-04:00       serviceAccountName: fake-server
    command.go:179: 2023-09-12T11:12:28-04:00       terminationGracePeriodSeconds: 0
    command.go:100: 2023-09-12T11:12:28-04:00 Running command kubectl with args [--namespace default wait --for=condition=available --timeout=5m deploy/fake-server]
    command.go:179: 2023-09-12T11:12:31-04:00 deployment.apps/fake-server condition met
    command.go:100: 2023-09-12T11:12:31-04:00 Running command kubectl with args [--namespace default get pod -l app=fake-server -o jsonpath="{.items[0].metadata.name}"]
    command.go:179: 2023-09-12T11:12:31-04:00 "fake-server-6c495cf7dc-j9ljn"
    observability_test.go:73: 2023-09-12T11:12:31-04:00 fake-server pod name:fake-server-6c495cf7dc-j9ljn
    tunnel.go:153: 2023-09-12T11:12:31-04:00 Creating a port forwarding tunnel for resource pod/fake-server-6c495cf7dc-j9ljn routing local port 57171 to remote port 443
TestObservabilityCloud 2023-09-12T11:12:31-04:00 client.go:33: Configuring kubectl using config file /Users/<omitted>/.kube/config with context 
    tunnel.go:185: 2023-09-12T11:12:31-04:00 Selected pod fake-server-6c495cf7dc-j9ljn to open port forward to
    tunnel.go:198: 2023-09-12T11:12:31-04:00 Using URL https://127.0.0.1:57123/api/v1/namespaces/default/pods/fake-server-6c495cf7dc-j9ljn/portforward to create portforward
    tunnel.go:247: 2023-09-12T11:12:31-04:00 Successfully created port forwarding tunnel
    observability_test.go:92: 2023-09-12T11:12:31-04:00 fake-server addr:localhost:57171
    observability_test.go:99: 2023-09-12T11:12:31-04:00 consul test token :7a3e7c8c-6c9e-4ec9-b6d2-d3f0c1ae2621
    command.go:100: 2023-09-12T11:12:31-04:00 Running command helm with args [list --output json]
    command.go:179: 2023-09-12T11:12:31-04:00 []
    command.go:100: 2023-09-12T11:12:31-04:00 Running command helm with args [install --timeout 15m --namespace default --version ../../../charts/consul --set client.extraConfig="{\"log_level\": \"TRACE\"}" --set connectInject.default=true --set connectInject.transparentProxy.defaultEnabled=false --set dns.enabled=false --set global.acls.manageSystemACLs=false --set global.cloud.apiHost.secretKey=apihost-sec-key --set global.cloud.apiHost.secretName=apihost-sec-name --set global.cloud.authUrl.secretKey=authurl-sec-key --set global.cloud.authUrl.secretName=authurl-sec-name --set global.cloud.clientId.secretKey=clientid-sec-key --set global.cloud.clientId.secretName=clientid-sec-name --set global.cloud.clientSecret.secretKey=client-sec-key --set global.cloud.clientSecret.secretName=client-sec-name --set global.cloud.enabled=true --set global.cloud.resourceId.secretKey=resource-sec-key --set global.cloud.resourceId.secretName=resource-sec-name --set global.cloud.scadaAddress.secretKey=scadaaddress-sec-key --set global.cloud.scadaAddress.secretName=scadaaddress-sec-name --set global.gossipEncryption.autoGenerate=false --set global.image=docker.io/achooo/consul01:17 --set global.imagePullPolicy=IfNotPresent --set global.logLevel=debug --set global.tls.enableAutoEncrypt=true --set global.tls.enabled=true --set server.extraConfig="{\"log_level\": \"TRACE\"}" --set server.extraEnvironmentVars.HCP_API_TLS=insecure --set server.extraEnvironmentVars.HCP_AUTH_TLS=insecure --set server.extraEnvironmentVars.HCP_SCADA_TLS=insecure --set server.replicas=1 --set telemetryCollector.cloud.clientId.secretKey=clientid-sec-key --set telemetryCollector.cloud.clientId.secretName=clientid-sec-name --set telemetryCollector.cloud.clientSecret.secretKey=client-sec-key --set telemetryCollector.cloud.clientSecret.secretName=client-sec-name --set telemetryCollector.enabled=true --set telemetryCollector.extraEnvironmentVars.HCP_API_TLS=insecure --set telemetryCollector.extraEnvironmentVars.HCP_AUTH_TLS=insecure --set telemetryCollector.extraEnvironmentVars.HCP_SCADA_TLS=insecure --set telemetryCollector.extraEnvironmentVars.OTLP_EXPORTER_TLS=insecure --set telemetryCollector.image=hashicorp/consul-telemetry-collector:0.0.1 test-hlzobx /<omitted>/Github/HashiCorp/consul-k8s/charts/consul]
    command.go:179: 2023-09-12T11:13:30-04:00 NAME: test-hlzobx
    command.go:179: 2023-09-12T11:13:30-04:00 LAST DEPLOYED: Tue Sep 12 11:12:31 2023
    command.go:179: 2023-09-12T11:13:30-04:00 NAMESPACE: default
    command.go:179: 2023-09-12T11:13:30-04:00 STATUS: deployed
    command.go:179: 2023-09-12T11:13:30-04:00 REVISION: 1
    command.go:179: 2023-09-12T11:13:30-04:00 NOTES:
    command.go:179: 2023-09-12T11:13:30-04:00 Thank you for installing HashiCorp Consul!
    command.go:179: 2023-09-12T11:13:30-04:00 
    command.go:179: 2023-09-12T11:13:30-04:00 Your release is named test-hlzobx.
    command.go:179: 2023-09-12T11:13:30-04:00 
    command.go:179: 2023-09-12T11:13:30-04:00 To learn more about the release, run:
    command.go:179: 2023-09-12T11:13:30-04:00 
    command.go:179: 2023-09-12T11:13:30-04:00   $ helm status test-hlzobx --namespace default
    command.go:179: 2023-09-12T11:13:30-04:00   $ helm get all test-hlzobx --namespace default
    command.go:179: 2023-09-12T11:13:30-04:00 
    command.go:179: 2023-09-12T11:13:30-04:00 Consul on Kubernetes Documentation:
    command.go:179: 2023-09-12T11:13:30-04:00 https://www.consul.io/docs/platform/k8s
    command.go:179: 2023-09-12T11:13:30-04:00 
    command.go:179: 2023-09-12T11:13:30-04:00 Consul on Kubernetes CLI Reference:
    command.go:179: 2023-09-12T11:13:30-04:00 https://www.consul.io/docs/k8s/k8s-cli
    observability_test.go:154: 2023-09-12T11:13:30-04:00 Waiting for pods with label "release=test-hlzobx" to be ready.
    observability_test.go:154: 2023-09-12T11:14:48-04:00 Finished waiting for pods to be ready.
    observability_test.go:156: 2023-09-12T11:14:48-04:00 creating static-server deployment
    command.go:100: 2023-09-12T11:14:48-04:00 Running command kubectl with args [--namespace default apply -k ../fixtures/bases/static-server]
    command.go:179: 2023-09-12T11:14:49-04:00 serviceaccount/static-server created
    command.go:179: 2023-09-12T11:14:49-04:00 rolebinding.rbac.authorization.k8s.io/static-server created
    command.go:179: 2023-09-12T11:14:49-04:00 rolebinding.rbac.authorization.k8s.io/static-server-openshift-anyuid created
    command.go:179: 2023-09-12T11:14:49-04:00 rolebinding.rbac.authorization.k8s.io/static-server-openshift-privileged created
    command.go:179: 2023-09-12T11:14:49-04:00 service/static-server created
    command.go:179: 2023-09-12T11:14:49-04:00 deployment.apps/static-server created
    command.go:100: 2023-09-12T11:14:49-04:00 Running command kubectl with args [--namespace default kustomize ../fixtures/bases/static-server]
    command.go:179: 2023-09-12T11:14:49-04:00 apiVersion: v1
    command.go:179: 2023-09-12T11:14:49-04:00 kind: ServiceAccount
    command.go:179: 2023-09-12T11:14:49-04:00 metadata:
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 ---
    command.go:179: 2023-09-12T11:14:49-04:00 apiVersion: rbac.authorization.k8s.io/v1
    command.go:179: 2023-09-12T11:14:49-04:00 kind: RoleBinding
    command.go:179: 2023-09-12T11:14:49-04:00 metadata:
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 roleRef:
    command.go:179: 2023-09-12T11:14:49-04:00   apiGroup: rbac.authorization.k8s.io
    command.go:179: 2023-09-12T11:14:49-04:00   kind: ClusterRole
    command.go:179: 2023-09-12T11:14:49-04:00   name: test-psp
    command.go:179: 2023-09-12T11:14:49-04:00 subjects:
    command.go:179: 2023-09-12T11:14:49-04:00 - kind: ServiceAccount
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 ---
    command.go:179: 2023-09-12T11:14:49-04:00 apiVersion: rbac.authorization.k8s.io/v1
    command.go:179: 2023-09-12T11:14:49-04:00 kind: RoleBinding
    command.go:179: 2023-09-12T11:14:49-04:00 metadata:
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server-openshift-anyuid
    command.go:179: 2023-09-12T11:14:49-04:00 roleRef:
    command.go:179: 2023-09-12T11:14:49-04:00   apiGroup: rbac.authorization.k8s.io
    command.go:179: 2023-09-12T11:14:49-04:00   kind: ClusterRole
    command.go:179: 2023-09-12T11:14:49-04:00   name: system:openshift:scc:anyuid
    command.go:179: 2023-09-12T11:14:49-04:00 subjects:
    command.go:179: 2023-09-12T11:14:49-04:00 - kind: ServiceAccount
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 ---
    command.go:179: 2023-09-12T11:14:49-04:00 apiVersion: rbac.authorization.k8s.io/v1
    command.go:179: 2023-09-12T11:14:49-04:00 kind: RoleBinding
    command.go:179: 2023-09-12T11:14:49-04:00 metadata:
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server-openshift-privileged
    command.go:179: 2023-09-12T11:14:49-04:00 roleRef:
    command.go:179: 2023-09-12T11:14:49-04:00   apiGroup: rbac.authorization.k8s.io
    command.go:179: 2023-09-12T11:14:49-04:00   kind: ClusterRole
    command.go:179: 2023-09-12T11:14:49-04:00   name: system:openshift:scc:privileged
    command.go:179: 2023-09-12T11:14:49-04:00 subjects:
    command.go:179: 2023-09-12T11:14:49-04:00 - kind: ServiceAccount
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 ---
    command.go:179: 2023-09-12T11:14:49-04:00 apiVersion: v1
    command.go:179: 2023-09-12T11:14:49-04:00 kind: Service
    command.go:179: 2023-09-12T11:14:49-04:00 metadata:
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 spec:
    command.go:179: 2023-09-12T11:14:49-04:00   ports:
    command.go:179: 2023-09-12T11:14:49-04:00   - name: http
    command.go:179: 2023-09-12T11:14:49-04:00     port: 80
    command.go:179: 2023-09-12T11:14:49-04:00     targetPort: 8080
    command.go:179: 2023-09-12T11:14:49-04:00   selector:
    command.go:179: 2023-09-12T11:14:49-04:00     app: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 ---
    command.go:179: 2023-09-12T11:14:49-04:00 apiVersion: apps/v1
    command.go:179: 2023-09-12T11:14:49-04:00 kind: Deployment
    command.go:179: 2023-09-12T11:14:49-04:00 metadata:
    command.go:179: 2023-09-12T11:14:49-04:00   name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00 spec:
    command.go:179: 2023-09-12T11:14:49-04:00   replicas: 1
    command.go:179: 2023-09-12T11:14:49-04:00   selector:
    command.go:179: 2023-09-12T11:14:49-04:00     matchLabels:
    command.go:179: 2023-09-12T11:14:49-04:00       app: static-server
    command.go:179: 2023-09-12T11:14:49-04:00   template:
    command.go:179: 2023-09-12T11:14:49-04:00     metadata:
    command.go:179: 2023-09-12T11:14:49-04:00       labels:
    command.go:179: 2023-09-12T11:14:49-04:00         app: static-server
    command.go:179: 2023-09-12T11:14:49-04:00       name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00     spec:
    command.go:179: 2023-09-12T11:14:49-04:00       containers:
    command.go:179: 2023-09-12T11:14:49-04:00       - args:
    command.go:179: 2023-09-12T11:14:49-04:00         - -text="hello world"
    command.go:179: 2023-09-12T11:14:49-04:00         - -listen=:8080
    command.go:179: 2023-09-12T11:14:49-04:00         image: docker.mirror.hashicorp.services/hashicorp/http-echo:alpine
    command.go:179: 2023-09-12T11:14:49-04:00         name: static-server
    command.go:179: 2023-09-12T11:14:49-04:00         ports:
    command.go:179: 2023-09-12T11:14:49-04:00         - containerPort: 8080
    command.go:179: 2023-09-12T11:14:49-04:00           name: http
    command.go:179: 2023-09-12T11:14:49-04:00       serviceAccountName: static-server
    command.go:179: 2023-09-12T11:14:49-04:00       terminationGracePeriodSeconds: 0
    command.go:100: 2023-09-12T11:14:49-04:00 Running command kubectl with args [--namespace default wait --for=condition=available --timeout=5m deploy/static-server]
    command.go:179: 2023-09-12T11:15:02-04:00 deployment.apps/static-server condition met
    observability_test.go:159: Finished deployment. Validating expected conditions now
=== RUN   TestObservabilityCloud/collectorExportsMetrics
=== RUN   TestObservabilityCloud/consulExportsMetrics
=== RUN   TestObservabilityCloud/consulPeriodicRefreshUpdateConfig
=== RUN   TestObservabilityCloud/consulPeriodicRefreshDisabled
=== NAME  TestObservabilityCloud
    helpers.go:101: 2023-09-12T11:16:02-04:00 cleaning up resources for TestObservabilityCloud
    command.go:100: 2023-09-12T11:16:02-04:00 Running command kubectl with args [--namespace default delete --timeout=120s --ignore-not-found -k ../fixtures/bases/static-server]
    command.go:179: 2023-09-12T11:16:02-04:00 serviceaccount "static-server" deleted
    command.go:179: 2023-09-12T11:16:02-04:00 rolebinding.rbac.authorization.k8s.io "static-server" deleted
    command.go:179: 2023-09-12T11:16:02-04:00 rolebinding.rbac.authorization.k8s.io "static-server-openshift-anyuid" deleted
    command.go:179: 2023-09-12T11:16:02-04:00 rolebinding.rbac.authorization.k8s.io "static-server-openshift-privileged" deleted
    command.go:179: 2023-09-12T11:16:02-04:00 service "static-server" deleted
    command.go:179: 2023-09-12T11:16:02-04:00 deployment.apps "static-server" deleted
    helpers.go:101: 2023-09-12T11:16:02-04:00 cleaning up resources for TestObservabilityCloud
    command.go:100: 2023-09-12T11:16:02-04:00 Running command helm with args [delete --keep-history --namespace default test-hlzobx]
    command.go:179: 2023-09-12T11:16:14-04:00 release "test-hlzobx" uninstalled
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
    command.go:100: 2023-09-12T11:16:15-04:00 Running command kubectl with args [--namespace default delete --timeout=120s --ignore-not-found -k ../fixtures/bases/cloud/hcp-mock]
    command.go:179: 2023-09-12T11:16:15-04:00 serviceaccount "fake-server" deleted
    command.go:179: 2023-09-12T11:16:15-04:00 service "fake-server" deleted
    command.go:179: 2023-09-12T11:16:15-04:00 deployment.apps "fake-server" deleted
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
    helpers.go:101: 2023-09-12T11:16:15-04:00 cleaning up resources for TestObservabilityCloud
--- PASS: TestObservabilityCloud (227.52s)
    --- PASS: TestObservabilityCloud/collectorExportsMetrics (0.02s)
    --- PASS: TestObservabilityCloud/consulExportsMetrics (0.01s)
    --- PASS: TestObservabilityCloud/consulPeriodicRefreshUpdateConfig (60.02s)
    --- PASS: TestObservabilityCloud/consulPeriodicRefreshDisabled (0.00s)
PASS
ok      github.com/hashicorp/consul-k8s/acceptance/tests/cloud  228.133s
```

How I expect reviewers to test this PR:


Checklist:
- [x] Tests added
- [x] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

Achooo commented on acceptance/tests/cloud/http_client.go:
```diff
@@ -0,0 +1,157 @@
+	req, err := http.NewRequest("POST", url, payloadBuf)
+	if err != nil {
+		fmt.Println("Error creating request:", err)
+		return errors.New("error creating modify_telemetry_config request")
```
will refactor + cleanup error handling

Achooo commented on acceptance/tests/cloud/fakeserver_client.go:
```diff
@@ -0,0 +1,164 @@
+)
+
+const (
+	// Validation are used in the validationBody to distinguish metrics for consul vs. collector during validation.
```
we could also make the fakeserver handle different paths : /validation/consul vs. validation/collector, which might be clearer but was more work on the fakeserver.

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -24,10 +18,6 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-type TokenResponse struct {
```
moved to `fakeserver_client.go`

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -54,47 +44,7 @@ var (
 	scadaAddressSecretKeyValue = "fake-server:443"
 )
 
-// The fake-server has a requestToken endpoint to retrieve the token.
```
moved to `fakeserver_client.go`

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -179,14 +130,12 @@ func TestBasicCloud(t *testing.T) {
 
 		"telemetryCollector.enabled":                   "true",
-		"telemetryCollector.image":                     cfg.ConsulCollectorImage,
+		"telemetryCollector.image":                     "hashicorp/consul-telemetry-collector:0.0.1",
```
hardcode this as test requires a collector to pass

curtbushko replied:
You shouldn't need to hardcode this as you are already passing it in as an argument to the test (like CI would)?  You should be able pull it through using the acceptance test framework flags. I think it is the only helm value that needs to be dynamic for these tests.

```
-consul-collector-image hashicorp/consul-telemetry-collector:0.0.1 
```

Achooo replied:
oh nvm it does throw an error when you don't do it 💯 

```
Received unexpected error:
                                error while running command: exit status 1; Error: INSTALLATION FAILED: execution error at (consul/templates/telemetry-collector-deployment.yaml:2:48): telemetryCollector.image must be set to enable consul-telemetry-collector
```

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -179,14 +130,12 @@ func TestBasicCloud(t *testing.T) {
 
 		"telemetryCollector.cloud.clientSecret.secretName": clientSecretName,
 		"telemetryCollector.cloud.clientSecret.secretKey":  clientSecretKey,
-		// Either we set the global.trustedCAs (make sure it's idented exactly) or we
```
@chapmanc deleted this wdyt?

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -196,33 +145,10 @@ func TestBasicCloud(t *testing.T) {
 		"server.extraEnvironmentVars.HCP_AUTH_TLS":  "insecure",
 		"server.extraEnvironmentVars.HCP_SCADA_TLS": "insecure",
-
-		// This is pregenerated CA used for testing. It can be replaced at any time and isn't
```
@chapmanc deleted this wdyt?

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -196,33 +145,10 @@ func TestBasicCloud(t *testing.T) {
 	if cfg.ConsulImage != "" {
 		helmValues["global.image"] = cfg.ConsulImage
 	}
-	if cfg.ConsulCollectorImage != "" {
```
We could keep this in case someone wants to pass in a collector image :shrug?

curtbushko replied:
Yes keep it, but see above for getting the value from the CLI.

Achooo replied:
:100 fixed above, don't need this line since it's set in the map above.

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -231,56 +157,78 @@ func TestBasicCloud(t *testing.T) {
+		})
 	}
-
-	// Perform the request
```
moved into `fakeserver_client.go`

Achooo commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -54,47 +44,7 @@ var (
-}
-
-func TestBasicCloud(t *testing.T) {
+func TestObservabilityCloud(t *testing.T) {
```
Renamed to `TestObservabilityCloud` since we're verifying metrics export from Collector + Consul mainly

curtbushko commented on acceptance/tests/cloud/observability_test.go:
```diff
@@ -231,56 +157,78 @@ func TestBasicCloud(t *testing.T) {
+
+			// Validate that exported metrics are correct using fakeserver's /validation endpoint, which records metric exports that occured.
+			// We need to use retry as we wait for Consul or the Collector to export metrics. This is the best we can do to avoid flakiness.
+			retry.RunWith(&retry.Timer{Timeout: tc.timeout, Wait: tc.wait}, t, func(r *retry.R) {
```
I definitely prefer retries over sleeps (in almost all cases).

curtbushko reviewed (changes requested):
Looks pretty good to me. 

Just one comment/request about hardcoding the collector image. I'll leave the logic validation up to you and Chappie. 😄

curtbushko reviewed (approved):
LGTM!

Achooo commented on acceptance/go.mod:
```diff
@@ -13,6 +13,8 @@ require (
 	github.com/hashicorp/serf v0.10.1
 	github.com/hashicorp/vault/api v1.8.3
 	github.com/stretchr/testify v1.8.3
+	go.opentelemetry.io/proto/otlp v1.0.0
```
Had to add otlp protos to do the metric validation

Achooo commented:
After discussing with @chapmanc, we decided to move the actual validation of metrics at the test itself instead of having an endpoint on the fakeserver, why?
- Logic in the fakeserver was becoming overly complex
- Extensibility (long term) in terms of accessing different metric level data to validate certain things. As functionality grows, and test cases will grow, we likely need custom validations and more control over this.

**SO, changes since the last review:**

In the test scenarios, for each test case, we now:
- Get recorded metrics:
  - Use `/records` endpoint instead to simply get back a record of metrics that were exported recently
  - Remove `/validation endpoint` and related code
- Validate metrics:
  -  New file `metrics_validation.go` that parses and verifies OTLP metrics  payload for the right metric names and labels.
  - Since metrics are OTLP protobufs, I added a new go module dependency

Achooo commented on acceptance/tests/cloud/metrics_validation.go:
This file contains helpers to validate metrics:
- Parse the OTLP payload
- Extract labels / metric names from OTLP payload

chapmanc reviewed (approved):
LGTM

Achooo commented:
Waiting to merge this as unrelated tests are failing CI. I see some PRs to get these fixed. Will merge after.

Achooo commented:
Apologies for the delay here, just rebased onto main, in the hopes of merging once again once CI all passes..
